README
/* gtkspell - a spell-checking addon for GTK's TextView widget
 * Copyright (c) 2002 Evan Martin
 * Copyright (c) 2012-2013 Sandro Mani
 */

Introduction
------------
GtkSpell is a library whihc provides MSWord/MacOSX-style highlighting of
misspelled words in a GtkTextView widget. Right-clicking a misspelled word pops
up a menu of suggested replacements.

GtkSpell depends on
 * GTK+3
 * Enchant
 * Gobject Introspection
 * GTK-Doc (optional)
 * GTK+2 (optional)

For more information, see http://sourceforge.net/projects/gtkspell/ .


Use
---
The easiest way to learn is to look at the provided example programs.
It's pretty simple.

Use gtk_spell_checker_new() and gtk_spell_checker_attach() to attach a new
GtkSpellChecker to a GtkTextView.

For convenience, gtk_spell_checker_new() is created as GInitiallyUnowned
and gtk_spell_checker_attach() will sink the floating reference.
gtk_spell_checker_attach() also connects the 'destroy' signal of the
passed-in GObject to g_object_unref() the GtkSpellChecker, so in the most
common use-case, you need not worry about cleaning it up, see
examples/simple.c.

However, if you want to detach and later reattach the GtkSpellChecker to a
(possibly different) GtkTextView, you must get a pointer to it with
gtk_spell_checker_get_from_text_view(), call g_object_ref() on the
resulting pointer, call gtk_spell_checker_detach() on it, call
gtk_spell_checker_attach() with the new GtkTextView, and finally call
g_object_unref() to release the reference that you took at the beginning.

Alternatively, you can sink the GtkSpellChecker immediately upon construction
with g_object_ref_sink(), in which case you will retain the ownership of the
GtkSpellChecker throughout the lifetime of the program, and you must remember
to call g_object_unref() when you don't need it any more, see
examples/advanced.c.

You don't need to keep a pointer to an attached GtkSpellChecker object around,
but you can retrieve it from a GtkTextView with
gtk_spell_checker_get_from_text_view() when necessary.

Once you have a pointer to a GtkSpellChecker, you can change the language it
uses with gtk_spell_checker_set_language().  Use gtk_spell_checker_detach()
to remove the GtkSpellChecker from its attached GtkTextView. Note that the
GtkSpellChecker will get destroyed when gtk_spell_checker_detach() is called
unless a reference is taken before via g_object_ref(). Finally,
gtk_spell_checker_recheck_all() is provided to re-spellcheck the entire buffer,
but it shouldn't be necessary.

An API reference can be found in the documentation. It the documentation was
not installed, gtkspell.h serves as an API reference.


Original Author
---------------
Evan Martin <martine@danga.com> - http://neugierig.org/

Previous Maintainers
--------------------
Nathan Fredrickson <nathan@silverorange.com>
Eduardo Pérez Ureta <eperez@users.sourceforge.net> - http://edpeur.blogspot.com/

Current Maintainers
------------------
Sandro Mani <manisandro@gmail.com>
Daniel Atallah

Thanks
------
Ari Pollak <ari@debian.org> - Debian packages.
Alex <hansolo@rochester.rr.com> - Redhat packages.


vim: tw=72


ChangeLog
version 3.0.10 (2018-06-30)
	* Port to enchant-2
	* Translation updates

version 3.0.9 (2016-08-08):
	* Add gtk_spell_checker_add_to_dictionary
	* Add gtk_spell_checker_ignore_word
	* Fix possible use-after-free in gtk_spell_checker_decode_language_code

version 3.0.8 (2016-04-03):
	* Fix cases where check_word is called on an empty range
	* Add gtk_spell_checker_check_word
	* Add gtk_spell_checker_get_suggestions
	* Translation updates

version 3.0.7 (2015-02-20):
	* Fix crash when building context menu if no spelling dictionary is active
	* Fix gtk_spell_checker_get_language_list docstring - g_list_free_full
	  needs to be used, not g_list_free
	* Translation updates

version 3.0.6 (2014-04-24):
	* Fix badly merged translations
	* Also recognize U+2019 as apostrophe
	* Translation updates (see po/ChangeLog)

version 3.0.5 (2014-04-19):
	* Fix incorrect handling of single quotes
	* Buildsystem cleanups and tweaks
	* Translation updates (see po/ChangeLog)

version 3.0.4 (2013-09-25):
	* Make decode-language-codes more robust
	* Don't use deprecated symbols when compiled against gtk3 >= 3.9
	* Translation updates (see po/ChangeLog)

version 3.0.3 (2013-06-04):
	* Add decode-language-codes property (to show more friendly dictionary
	  names in the context menu), implemented using iso-codes
	* Add gtk_spell_checker_decode_language_code method
	* Use g_win32_get_package_installation_directory_of_module on win32
	* Add -no-undefined when building on for win32
	* Use $(foo) instead of @foo@ in Makefiles
	* Fix GTK_SPELL_CFLAGS -> GTK_SPELL3_CFLAGS typo
	* Translation updates (see po/ChangeLog)

version 3.0.2 (2013-03-05):
	* Added vala bindings (thanks to Evan Nemerson)
	* Added some documentation on language bindings

version 3.0.1 (2013-02-07):
	* Actually declare GtkSpellChecker as a G_TYPE_INITIALLY_UNOWNED
	* Really silently exit if detach is called when not attached
	* Add gtk_spell_error_get_type
	* Use PIMPL idiom for private members of GtkSpellChecker
	* Update license file with correct FSF address, add license headers to files
	* New and updated translations

version 3.0.0 (2012-11-12):
	* New major version not API or ABI compatible with gtkspell 2.x
	  The tutorial and examples have been updated with lots of details.
	* Add GTK+ 3 Support (required by default)
	 * A GTK+ 2 compatible version can be built using --enable-gtk2
	* Update to be GObject based
	* Add introspection support
	* Add "language-changed" signal
	* Add gtk_spell_checker_get_language function

version 2.0.16 (2009-10-22):
	* Additional and updated translations (see po/ChangeLog).
	* Fix Bug #2431005 Language codes that contain "C" (e.g. "cs_CZ")
	  defaulted to "en" (Blondak)
	* Fix Bug #2715972 Remove an extra comma at the end of the GtkSpellError
	  enum definition.
	* Fix Bug #2526949 Sort languages list submenu.
	* Currently selected language appears with radio button selection in
	  Language selector sub-menu (Feature request #2526959)
	* Fix Bug #2642198 Enchant Broker should be a singleton.
	  This will cause significantly lower memory usage when using multiple
	  GtkSpell instances.
	* Fix Bug #2875116 Support changing of the GtkTextBuffer in the attached
	  GtkTextView. (Colomban Wendling)

version 2.0.15 (2008-12-01):
	* Additional translations (see po/ChangeLog).
	* Add missing G_BEGIN_DECLS/G_END_DECLS for C++ support
	* Fix feature request #2101451 by removing the sub-menu
	  of the context menu as any other in-line spell checker does.
	* Fix feature request #1523881, #1643622 and #2054637
	  by adding a Language selector sub-menu.
	* Fix feature request #1362304 by fixing docs/Makefile.am
	  to install gtkspell.devhelp so that it shows in devhelp

version 2.0.14 (2008-08-15):
	* Additional translations (see po/ChangeLog).
	* Changed bug reporting address at configure.ac:AC_INIT
	* undo/redo in the calling application (Liviu Andronic)
	* gtkspell_get_suggestions_menu() (Liviu Andronic)

version 2.0.13 (2008-05-20):
	* Additional translations (see po/ChangeLog).
	* Enchant support (support for pspell/aspell dropped) (Dom Lachowicz)

version 2.0.12 (2008-05-09):
	* Make debug, quiet and build_suggestion_menu static in gtkspell.c
	  to prevent exporting them.
	* One additional translation (see po/ChangeLog).
	* Move spell-checking library from Libs to Libs.private in
	  gtkspell-2.0.pc 
	* Update autogen.sh to use intltool
	* Fix warnings with -Wstrict-prototypes by Richard Laager

version 2.0.11 (2005-05-27):
	* Re-enable apostrophe word-breaking workarounds (GNOME bug #97545).
	* Additional translations (see po/ChangeLog).

version 2.0.10 (2005-02-27):
	* Re-release 2.0.9 with debug mode turned off.
	* One additional translation (see po/ChangeLog).

version 2.0.9 (2005-02-22):
	* Enable building outside of srcdir (Mike Castle)
	* Use bind_textdomain_codeset to set UTF-8 encoding for all translations.
	* Do not delete the GtkTextTag from the tag table (needed for shared
	  tag tables).
	* Disable word-breaking workarounds when Pango >= 1.8.0 (GNOME bug #97545).
	* Many additional translations (see po/ChangeLog).

version 2.0.8 (2004-10-08):
	* Fix how the GtkTextTag is created to work with shared tag tables.
	* Additional translation (see po/ChangeLog).

version 2.0.7 (2004-08-22):
	* Activating the menu through a keybinding now properly checks the
	  word currently under the cursor.
	* Prevent the trampling on an application's textdomain (Stu Tomlinson).
	* Additional translations (see po/ChangeLog).

version 2.0.6 (2004-06-26):
	* Use new PANGO_UNDERLINE_ERROR style when available to highlight
	  misspelled words with squiggly red underline.
	* Internationalize with GNU Gettext. All five strings now translatable.
	  Thanks to Adriano Del Vigna de Almeida <katmandu@fs.inf.br>
	* New po/ChangeLog to track translation additions/updates.

version 2.0.5 (2004-01-26):
	* Don't check numeric words.  Combined with word breaking changes 
	  recently committed to Pango, this will handle all letter-number

autogen.sh
#!/bin/sh
# Run this to generate all the initial makefiles, etc.

set -o errexit

srcdir="$(dirname "$(readlink -f "$0")")"

(test -f "$srcdir"/configure.ac) || {
  echo -n "**Error**: Directory "\`$srcdir\'" does not look like the"
  echo " top-level package directory"
  echo
  exit 1
}

(intltoolize --version) < /dev/null > /dev/null 2>&1 || {
  echo "**Error**: You must have \`intltool' installed."
  echo "You can get it from:"
  echo "  https://launchpad.net/intltool"
  echo
  exit 1
}

(libtool --version) < /dev/null > /dev/null 2>&1 || {
  echo "**Error**: You must have \`libtool' installed."
  echo "You can get it from: ftp://ftp.gnu.org/pub/gnu/"
  echo
  exit 1
}

(autoreconf --version) < /dev/null > /dev/null 2>&1 || {
  echo "**Error**: You must have \`autoconf' installed."
  echo "Download the appropriate package for your distribution,"
  echo "or get the source tarball at ftp://ftp.gnu.org/pub/gnu/"
  echo
  exit 1
}

(aclocal --version) < /dev/null > /dev/null 2>&1 || {
  echo "**Error**: You must have \`automake' installed."
  echo "Download the appropriate package for your distribution,"
  echo "or get the source tarball at ftp://ftp.gnu.org/pub/gnu/"
  echo
  exit 1
}

(stat "$(aclocal --print-ac-dir)/ax_cxx_compile_stdcxx_11.m4") < /dev/null > /dev/null 2>&1 || {
  echo "**Error**: You must have the \`ax_cxx_compile_stdcxx_11.m4' macro installed-"
  echo "Download the appropriate package for your distribution,"
  echo "or get it from http://mirror.switch.ch/ftp/mirror/gnu/autoconf-archive/"
  echo
  exit 1
}


(
  cd "$srcdir"
  mkdir -p build
  echo "Running intltoolize..."
  intltoolize --force --copy --automake || exit 1
)


echo "Running gtkdocize..."
GTKDOCIZE=`which gtkdocize`
if test -z $GTKDOCIZE; then
  echo "gtkdocize not found, skipping"
else
  gtkdocize --srcdir "$srcdir" || exit 1
fi


echo "Running autoreconf..."
autoreconf -fiv "$srcdir" || exit 1


if test x$NOCONFIGURE = x; then
  echo Running $srcdir/configure "$@" ...
  "$srcdir"/configure "$@" \
  && echo Now type \`make\' to compile. || exit 1
else
  echo Skipping configure process.
fi


examples/simple.c
/* vim: set ts=4 sw=4 wm=5 : */

#include <gtk/gtk.h>
#include <gtkspell/gtkspell.h>

int
main (int argc, char* argv[])
{
  GtkWidget *win, *box, *scroll, *view;
  GError *error = NULL;
  GtkSpellChecker* spell;
  char *errortext = NULL;

  gtk_init (&argc, &argv);

  view = gtk_text_view_new ();
  gtk_text_view_set_wrap_mode (GTK_TEXT_VIEW (view), GTK_WRAP_WORD);

  spell = gtk_spell_checker_new ();
  if (!gtk_spell_checker_set_language (spell, "en_US", &error))
    {
      g_print ("gtkspell error: %s\n", error->message);
      errortext = g_strdup_printf ("GtkSpell was unable to initialize.\n"
                                   "%s", error->message);
      g_error_free (error);
    }
  gtk_spell_checker_attach (spell, GTK_TEXT_VIEW (view));

  scroll = gtk_scrolled_window_new (NULL, NULL);
  gtk_scrolled_window_set_policy (GTK_SCROLLED_WINDOW (scroll),
                                  GTK_POLICY_AUTOMATIC,
                                  GTK_POLICY_AUTOMATIC);
  gtk_scrolled_window_set_shadow_type (GTK_SCROLLED_WINDOW (scroll),
                                       GTK_SHADOW_IN);
  gtk_container_add (GTK_CONTAINER (scroll), view);

  box = gtk_box_new (GTK_ORIENTATION_VERTICAL, 5);
  if (errortext)
    {
      gtk_box_pack_start (GTK_BOX (box), gtk_label_new (errortext),
                          FALSE, FALSE, 0);
      g_free (errortext);
    }
  else
    {
      gtk_box_pack_start (GTK_BOX (box),
                          gtk_label_new ("Type some text into the text box.\n"
                     "Try misspelling some words.  Then right-click on them."),
                          FALSE, FALSE, 0);
    }
  gtk_box_pack_start (GTK_BOX (box), scroll, TRUE, TRUE, 0);
  gtk_widget_show_all (box);

  win = gtk_window_new (GTK_WINDOW_TOPLEVEL);
  gtk_window_set_default_size (GTK_WINDOW (win), 400, 300);
  gtk_window_set_title (GTK_WINDOW (win), "Simple GtkSpell Demonstration");
  gtk_container_set_border_width (GTK_CONTAINER (win), 10);
  g_signal_connect (win, "delete-event", G_CALLBACK (gtk_main_quit), NULL);
  gtk_container_add (GTK_CONTAINER (win), box);

  gtk_widget_show (win);
  gtk_main ();

  return 0;
}


examples/advanced.c
/* vim: set ts=4 sw=4 wm=5 : */

/* This example demonstrates detaching and reattaching GtkSpell. */

#include <gtk/gtk.h>
#include <gtkspell/gtkspell.h>

GtkWidget *window, *attached, *view;
GtkSpellChecker *spell;

static void
report_gtkspell_error (const char *err)
{
  GtkWidget *dlg;
  dlg = gtk_message_dialog_new (GTK_WINDOW (window),
                                GTK_DIALOG_DESTROY_WITH_PARENT,
                                GTK_MESSAGE_ERROR,
                                GTK_BUTTONS_CLOSE,
                                "GtkSpell error: %s", err);
  gtk_dialog_run (GTK_DIALOG (dlg));
  gtk_widget_destroy (dlg);
}

static void
attach_cb ()
{
  if (gtk_toggle_button_get_active (GTK_TOGGLE_BUTTON (attached)))
    {
      gtk_spell_checker_attach (spell, GTK_TEXT_VIEW (view));
    }
  else
    {
      gtk_spell_checker_detach (spell);
    }
}

int
main (int argc, char* argv[])
{
  GtkWidget *box, *hbox, *scroll;
  GError *error = NULL;

  gtk_init (&argc, &argv);

  spell = gtk_spell_checker_new ();
  g_object_ref_sink (spell);
  if (!gtk_spell_checker_set_language (spell, "en_US", &error))
    {
      report_gtkspell_error (error->message);
      g_error_free(error);
    }

  window = gtk_window_new (GTK_WINDOW_TOPLEVEL);

  view = gtk_text_view_new ();
  gtk_text_view_set_wrap_mode (GTK_TEXT_VIEW (view), GTK_WRAP_WORD);

  scroll = gtk_scrolled_window_new (NULL, NULL);
  gtk_scrolled_window_set_policy (GTK_SCROLLED_WINDOW (scroll),
                                  GTK_POLICY_AUTOMATIC,
                                  GTK_POLICY_AUTOMATIC);
  gtk_scrolled_window_set_shadow_type (GTK_SCROLLED_WINDOW (scroll),
                                       GTK_SHADOW_IN);
  gtk_container_add (GTK_CONTAINER (scroll), view);

  hbox = gtk_box_new (GTK_ORIENTATION_HORIZONTAL, 5);
  attached = gtk_toggle_button_new_with_label ("Attached");
  g_signal_connect (attached, "toggled", G_CALLBACK (attach_cb), NULL);
  gtk_box_pack_start (GTK_BOX (hbox), attached, FALSE, FALSE, 0);

  box = gtk_box_new (GTK_ORIENTATION_VERTICAL, 5);
  gtk_box_pack_start (GTK_BOX (box), scroll, TRUE, TRUE, 0);
  gtk_box_pack_start (GTK_BOX (box), hbox, FALSE, FALSE, 0);
  gtk_widget_show_all (box);

  gtk_window_set_default_size (GTK_WINDOW (window), 400, 300);
  gtk_window_set_title (GTK_WINDOW (window),
                        "\"Advanced\" GtkSpell Demonstration");
  gtk_container_set_border_width (GTK_CONTAINER (window), 10);
  g_signal_connect (G_OBJECT (window), "delete-event",
                    G_CALLBACK (gtk_main_quit), NULL);
  gtk_container_add (GTK_CONTAINER (window), box);

  gtk_widget_show (window);

  gtk_toggle_button_set_active (GTK_TOGGLE_BUTTON (attached), TRUE);

  gtk_main ();
  g_object_unref (spell);

  return 0;
}


gtkspell/gtkspell.h
/* gtkspell - a spell-checking addon for GTK's TextView widget
 * Copyright (c) 2002 Evan Martin
 * Copyright (c) 2012-2013 Sandro Mani
 *
 *    This program is free software; you can redistribute it and/or modify
 *    it under the terms of the GNU General Public License as published by
 *    the Free Software Foundation; either version 2 of the License, or
 *    (at your option) any later version.
 *
 *    This program is distributed in the hope that it will be useful,
 *    but WITHOUT ANY WARRANTY; without even the implied warranty of
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *    GNU General Public License for more details.
 *
 *    You should have received a copy of the GNU General Public License along
 *    with this program; if not, write to the Free Software Foundation, Inc.,
 *    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

/* vim: set ts=4 sw=4 wm=5 : */

#ifndef GTK_SPELL_H
#define GTK_SPELL_H

#include <glib-object.h>
#include <gtk/gtk.h>

G_BEGIN_DECLS

/**
* GTK_SPELL_ERROR:
*
* Error domain used for gtkspell operations. Indicates that the error code
* will be in the #GtkSpellError enumeration.
*/
#define GTK_SPELL_ERROR (gtk_spell_error_quark())
GQuark gtk_spell_error_quark(void);

/**
* GtkSpellError:
* @GTK_SPELL_ERROR_BACKEND: Error code for backend errors
*
* Error codes used for GtkSpell errors.
*
**/
typedef enum {
  GTK_SPELL_ERROR_BACKEND
} GtkSpellError;

/**
 * GTK_SPELL_TYPE_ERROR:
 *
 * The GTK_SPELL_ERROR type.
 */
#define GTK_SPELL_TYPE_ERROR (gtk_spell_error_get_type ())
GType gtk_spell_error_get_type (void) G_GNUC_CONST;

#define GTK_SPELL_TYPE_CHECKER            (gtk_spell_checker_get_type ())
#define GTK_SPELL_CHECKER(obj)            (G_TYPE_CHECK_INSTANCE_CAST ((obj), GTK_SPELL_TYPE_CHECKER, GtkSpellChecker))
#define GTK_SPELL_IS_CHECKER(obj)         (G_TYPE_CHECK_INSTANCE_TYPE ((obj), GTK_SPELL_TYPE_CHECKER))
#define GTK_SPELL_CHECKER_CLASS(klass)    (G_TYPE_CHECK_CLASS_CAST ((klass), GTK_SPELL_TYPE_CHECKER, GtkSpellCheckerClass))
#define GTK_SPELL_IS_CHECKER_CLASS(klass) (G_TYPE_CHECK_CLASS_TYPE ((klass), GTK_SPELL_TYPE_CHECKER))
#define GTK_SPELL_CHECKER_GET_CLASS(obj)  (G_TYPE_INSTANCE_GET_CLASS ((obj), GTK_SPELL_TYPE_CHECKER, GtkSpellCheckerClass))

/**
 * GtkSpellChecker:
 *
 * The #GtkSpellChecker struct contains only private fields.
 */
typedef struct _GtkSpellChecker        GtkSpellChecker;
typedef struct _GtkSpellCheckerPrivate GtkSpellCheckerPrivate;
struct _GtkSpellChecker
{
  GInitiallyUnowned parent_instance;

  /*< private >*/
  GtkSpellCheckerPrivate *priv;
};

typedef struct _GtkSpellCheckerClass   GtkSpellCheckerClass;
struct _GtkSpellCheckerClass
{
  GInitiallyUnownedClass parent_class;

  /*< public >*/
  void (*language_changed) (GtkSpellChecker *spell, const gchar *new_lang);
};


GType            gtk_spell_checker_get_type             (void) G_GNUC_CONST;
GtkSpellChecker *gtk_spell_checker_new                  (void);
gboolean         gtk_spell_checker_attach               (GtkSpellChecker *spell,
                                                         GtkTextView   *view);
GtkSpellChecker *gtk_spell_checker_get_from_text_view   (GtkTextView   *view);
void             gtk_spell_checker_detach               (GtkSpellChecker *spell);
GList           *gtk_spell_checker_get_suggestions      (GtkSpellChecker *spell,
                                                         const gchar* word);
GtkWidget       *gtk_spell_checker_get_suggestions_menu (GtkSpellChecker *spell,
                                                         GtkTextIter   *iter);
gboolean         gtk_spell_checker_set_language         (GtkSpellChecker *spell,
                                                         const gchar   *lang,
                                                         GError       **error);
const gchar     *gtk_spell_checker_get_language         (GtkSpellChecker *spell);
GList           *gtk_spell_checker_get_language_list    (void);
gchar           *gtk_spell_checker_decode_language_code (const gchar *lang);
gboolean         gtk_spell_checker_check_word           (GtkSpellChecker *spell,
                                                         const gchar *word);
void             gtk_spell_checker_recheck_all          (GtkSpellChecker *spell);

G_END_DECLS

#endif /* GTK_SPELL_H */


gtkspell/gtkspell-win32.h
/* gtkspell - a spell-checking addon for GTK's TextView widget
 * Copyright (c) 2013 Sandro Mani
 *
 * Based on gtkhtml-editor-spell-language.c code which is
 * Copyright (C) 2008 Novell, Inc.
 *
 *    This program is free software; you can redistribute it and/or modify
 *    it under the terms of the GNU General Public License as published by
 *    the Free Software Foundation; either version 2 of the License, or
 *    (at your option) any later version.
 *
 *    This program is distributed in the hope that it will be useful,
 *    but WITHOUT ANY WARRANTY; without even the implied warranty of
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *    GNU General Public License for more details.
 *
 *    You should have received a copy of the GNU General Public License along
 *    with this program; if not, write to the Free Software Foundation, Inc.,
 *    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

/* vim: set ts=4 sw=4 wm=5 : */

#ifndef GTK_SPELL_WIN32_H
#define GTK_SPELL_WIN32_H

#include "../config.h"
#include <glib.h>

G_BEGIN_DECLS

const gchar *get_win32_prefix    (void);
const gchar *get_win32_localedir (void);

G_END_DECLS

#undef PACKAGE_LOCALE_DIR
#define PACKAGE_LOCALE_DIR get_win32_localedir ()

#ifdef HAVE_ISO_CODES

#undef ISO_CODES_PREFIX
#define ISO_CODES_PREFIX get_win32_prefix ()

#undef ISOCODESLOCALEDIR
#define ISOCODESLOCALEDIR get_win32_localedir ()

#endif /* HAVE_ISO_CODES */

#endif /* GTK_SPELL_WIN32_H */


gtkspell/gtkspell-codetable.c
/* gtkspell - a spell-checking addon for GTK's TextView widget
 * Copyright (c) 2013 Sandro Mani
 *
 * Based on gtkhtml-editor-spell-language.c code which is
 * Copyright (C) 2008 Novell, Inc.
 *
 *    This program is free software; you can redistribute it and/or modify
 *    it under the terms of the GNU General Public License as published by
 *    the Free Software Foundation; either version 2 of the License, or
 *    (at your option) any later version.
 *
 *    This program is distributed in the hope that it will be useful,
 *    but WITHOUT ANY WARRANTY; without even the implied warranty of
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *    GNU General Public License for more details.
 *
 *    You should have received a copy of the GNU General Public License along
 *    with this program; if not, write to the Free Software Foundation, Inc.,
 *    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

/* vim: set ts=4 sw=4 wm=5 : */

#include "gtkspell-codetable.h"
#include "../config.h"
#include <libintl.h>
#include <string.h>

#ifdef G_OS_WIN32
#include "gtkspell-win32.h"
#endif

#define ISO_639_DOMAIN	"iso_639"
#define ISO_3166_DOMAIN	"iso_3166"

static GHashTable *iso_639_table = NULL;
static GHashTable *iso_3166_table = NULL;

static void
iso_639_start_element (GMarkupParseContext *context,
                       const gchar *element_name,
                       const gchar **attribute_names,
                       const gchar **attribute_values,
                       gpointer data,
                       GError **error)
{
  GHashTable *hash_table = data;
  const gchar *name = NULL;
  const gchar *code = NULL;
  int i;

  if (strcmp (element_name, "iso_639_entry") != 0)
    return;

  for (i = 0; attribute_names[i] != NULL; i++)
    {
      if (strcmp (attribute_names[i], "name") == 0)
        name = attribute_values[i];
      else if (strcmp (attribute_names[i], "iso_639_1_code") == 0)
        code = attribute_values[i];
    }

  if (code != NULL && *code != '\0' && name != NULL && *name != '\0')
    g_hash_table_insert (hash_table, g_strdup (code),
                         g_strdup (dgettext (ISO_639_DOMAIN, name)));
}

static void
iso_3166_start_element (GMarkupParseContext *context,
                        const gchar *element_name,
                        const gchar **attribute_names,
                        const gchar **attribute_values,
                        gpointer data,
                        GError **error)
{
  GHashTable *hash_table = data;
  const gchar *name = NULL;
  const gchar *code = NULL;
  int i;

  if (strcmp (element_name, "iso_3166_entry") != 0)
    return;

  for (i = 0; attribute_names[i] != NULL; i++)
    {
      if (strcmp (attribute_names[i], "name") == 0)
        name = attribute_values[i];
      else if (strcmp (attribute_names[i], "alpha_2_code") == 0)
        code = attribute_values[i];
    }

  if (code != NULL && *code != '\0' && name != NULL && *name != '\0')
    g_hash_table_insert (hash_table, g_strdup (code),
                         g_strdup (dgettext (ISO_3166_DOMAIN, name)));
}

static void
iso_codes_parse (const GMarkupParser *parser,
                 const gchar *basename,
                 GHashTable *hash_table)
{
  GMappedFile *mapped_file;
  gchar *filename;
  GError *error = NULL;

  filename = g_build_filename (ISO_CODES_PREFIX, "share", "xml", "iso-codes",
                               basename, NULL);
  mapped_file = g_mapped_file_new (filename, FALSE, &error);
  g_free (filename);

  if (mapped_file != NULL)
    {
      GMarkupParseContext *context;
      const gchar *contents;
      gsize length;

      context = g_markup_parse_context_new (parser, 0, hash_table, NULL);
      contents = g_mapped_file_get_contents (mapped_file);
      length = g_mapped_file_get_length (mapped_file);
      g_markup_parse_context_parse (context, contents, length, &error);
      g_markup_parse_context_free (context);
      g_mapped_file_unref (mapped_file);
    }

  if (error != NULL)
    {
      g_warning ("%s: %s", basename, error->message);
      g_error_free (error);
    }
}

/**
 * codetable_init:
 *
 * Initializes the code table.
 */
void
codetable_init (void)
{
  GMarkupParser iso_639_parser = {
    iso_639_start_element, NULL, NULL, NULL, NULL
  };

  GMarkupParser iso_3166_parser = {
    iso_3166_start_element, NULL, NULL, NULL, NULL
  };

  g_return_if_fail (iso_639_table == NULL);
  g_return_if_fail (iso_3166_table == NULL);

#ifdef ENABLE_NLS
  bindtextdomain (ISO_639_DOMAIN, ISO_CODES_LOCALEDIR);
  bind_textdomain_codeset (ISO_639_DOMAIN, "UTF-8");

  bindtextdomain (ISO_3166_DOMAIN, ISO_CODES_LOCALEDIR);
  bind_textdomain_codeset (ISO_3166_DOMAIN, "UTF-8");
#endif

  iso_639_table = g_hash_table_new_full (g_str_hash, g_str_equal,
                             (GDestroyNotify) g_free, (GDestroyNotify) g_free);
  iso_3166_table = g_hash_table_new_full (g_str_hash, g_str_equal,
                             (GDestroyNotify) g_free, (GDestroyNotify) g_free);

  iso_codes_parse (&iso_639_parser, "iso_639.xml", iso_639_table);
  iso_codes_parse (&iso_3166_parser, "iso_3166.xml", iso_3166_table);
}

/**
 * codetable_free:
 *
 * Frees the code table.
 */
void
codetable_free (void)
{
  g_return_if_fail (iso_639_table != NULL);
  g_return_if_fail (iso_3166_table != NULL);

  g_hash_table_unref (iso_639_table);
  g_hash_table_unref (iso_3166_table);

  iso_639_table = NULL;
  iso_3166_table = NULL;
}

/**
 * codetable_lookup:
 * @language_code: A language code (i.e. "en_US")
 * @language_name: (out) (transfer none): Pointer to the name of the language.
 * This pointer is owned by the code table and must not be freed.
 * @country_name: (out) (transfer none): Pointer to the name of the country.
 * This pointer is owned by the code table and must not be freed.
 *
 * Looks up the language and country name for the specified language code.
 * If no matching entries are found, language_name and country_name will
 * simply contain the parts of the language code (i.e. "en" and "US").
 */
void
codetable_lookup (const gchar *language_code, const gchar **language_name, const gchar** country_name)
{
  gchar **parts;

  g_return_if_fail (iso_639_table != NULL);
  g_return_if_fail (iso_3166_table != NULL);

  /* Split language code into parts. */
  parts = g_strsplit (language_code, "_", 2);

  g_return_if_fail (parts != NULL);

  if (g_strv_length (parts) != 2)
    {
      g_strfreev (parts);
      g_return_if_reached ();
	}

  *language_name = g_hash_table_lookup (iso_639_table, parts[0]);
  if (*language_name == NULL)
    {
      g_hash_table_insert (iso_639_table, g_strdup (parts[0]),
                           g_strdup (parts[0]));
      *language_name = g_hash_table_lookup (iso_639_table, parts[0]);
    }

  *country_name = g_hash_table_lookup (iso_3166_table, parts[1]);
  if (*country_name == NULL)
    {
      g_hash_table_insert (iso_3166_table, g_strdup (parts[1]),
                           g_strdup (parts[1]));
      *country_name = g_hash_table_lookup (iso_3166_table, parts[1]);
    }

  g_strfreev (parts);
}
